### PR TITLE
ethcore-db crate

### PR DESCRIPTION
**Pull request by NikVolf**
dedicated service for database exposed via ipc

client for database service (which is sharing `DatabaseService` trait for key-value operations) can be generated just like that:

```
let client = nanoipc::init_duplex_client::<DatabaseClient<_>>("~/.parity/db.ipc").unwrap();
client.open(DatabaseConfig { prefix_size: None }, "~/.parity/db").unwrap();
```

see `mod client_tests`

also contains some cryptic rpc codegen tricks to allow passing `&[u8]` in method signature

### CODE REVIEW

**Comment by arkpar** [comment]
Does this add a new binary?
Any performance degradation when doing a full sync?

**Comment by NikVolf** [comment]
@arkpar 
it's just a crate so far
not used in the main binary (next pr)

File: db/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+homepage = "http://ethcore.io"
+license = "GPL-3.0"
+name = "ethcore-db"
+version = "1.2.0"
+authors = ["Ethcore <admin@ethcore.io>"]
+build = "build.rs"
+
+[build-dependencies]
+syntex = "*"
+"ethcore-ipc-codegen" = { path = "../ipc/codegen" }
```
Comment (arkpar) [inline db/Cargo.toml]: Remove quotes for consistency

File: db/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+[dependencies]
+ethcore-util = { path = "../util" }
+clippy = { version = "0.0.67", optional = true}
+ethcore-devtools = { path = "../devtools" }
+"ethcore-ipc" = { path = "../ipc/rpc" }
```
Comment (arkpar) [inline db/Cargo.toml]: Remove quotes for consistency

File: db/src/database.rs [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+
+impl From<String> for Error {
+	fn from(s: String) -> Error {
+		Error::RocksDb(s)
+	}
+}
+
+pub struct Database {
+	db: RwLock<Option<DB>>,
+	is_open: RwLock<bool>,
```
Comment (arkpar) [inline db/src/database.rs]: Can this be `AtomicBool`? Or even better just check that `db.is_some()` ?

File: db/src/database.rs [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+
+		*is_open = true;
+		Ok(())
+	}
+
+	fn close(&self) -> Result<(), Error> {
+		let mut is_open = self.is_open.write().unwrap();
+		if !*is_open { return Err(Error::IsClosed); }
+
+		// TODO: wait for transactions to expire/close and destroy self.db?
```
Comment (arkpar) [inline db/src/database.rs]: Please add an assert for that at least

**Pull request merged**